### DISCUSSION TITLE
Typings for QueryClient functions inferred from query or mutation keys

### DISCUSSION THREAD

**Discussion by roblabat** [TanStack/query#6105] [Ideas]
Hey, I would like to share here an Idea of a way to implement type inferrence of data from the query key or mutation key in the QueryClient functions getQueryData, setQueryData...

This Idea comes from the implementation of types inference for the functionality provide/inject in Vue.

The Idea is to define the key as an exported const with a typing hosting the type information of the query and then inferring the getQueryData type from the type of the key if possible:

Usage would look something like that:

``` ts
// key file

import { QueryKey } from '@tanstack/vue-query';

interface Todo {
  text: string;
  finished: boolean;
}

export const TodosKey = ['todos'] as QueryKey<Todo[]>;

// Query file

const result = useQuery({
  queryKey: TodosKey,
  // here fetchTodos needs to return a promise of Todo[]
  queryFn: fetchTodos 
});

// Mutation file

useMutation({
  mutationFn: updateTodo,
  onMutate: async (newTodo) => {
    await queryClient.cancelQueries({ queryKey: TodosKey })

    // previousTodos is inferred as Todo[]
    const previousTodos = queryClient.getQueryData(TodosKey)

    // newTodos is checked to be of type Todo[]
    queryClient.setQueryData(TodosKey, (old) => [...old, newTodo])

    return { previousTodos }
  },
  onError: (err, newTodo, context) => {
    // context.previousTodos is checked to be of type Todo[]
    queryClient.setQueryData(TodosKey, context.previousTodos)
  },
  onSettled: () => {
    queryClient.invalidateQueries({ queryKey: TodosKey })
  },
})
```

Of course with this solution using a key directly as ['todos'] without using the shared TodosKey will not infer typings as currently but it would offer a way for users to type there queries as much as possible.

Maybe other currently untyped methods could also benefits from this patern.

My example is written for Vue as I'm currently working with the Vue package  of TanstackQuery but this pattern could be applied to all implementations.

If this solution is considered a good solution I would be able to take a bit of time to make a PR implementing this.

Waiting for your feedbacks :)

Ps: see [https://vuejs.org/guide/typescript/composition-api.html#typing-provide-inject](https://vuejs.org/guide/typescript/composition-api.html#typing-provide-inject) for my inspiration source in Vue provide/inject mechanism

**Comment by TkDodo** [comment]
the idea that I've had for v5 is the following:

we have a new function `queryOptions`, that's already done and type-safe. With it, you can do:

```
const todoOptions = queryOptions({
  queryKey: ['todos'],
  queryFn: fetchTodos
})
```

this one is fully typed just like `useQuery`, but it only creates the options that you can then pass to multiple functions:

```
useQuery(todoOptions)
queryClient.prefetchQuery(todoOptions)
useQueries({
  queries: [todoOptions]
})
```

so, its an abstraction to share the options needed to run a query. At runtime, it's just an identity function: it returns the options you pass in:

https://github.com/TanStack/query/blob/3f057a24778c15c55665d11fc8415200a20e4840/packages/react-query/src/queryOptions.ts#L56-L58

50 lines of types, 1 line of runtime code 😅 

So with that, I'd expect most users who want the best type safety to define their query options with that function. Now what if we'd take that a step further and we would allow passing `queryOptions` directly to the `QueryClient`:

```
new QueryClient({
  queryOptions: [todoOptions, postOptions]
})
```

This would basically be the upfront definition that we'd need to make `queryClient.getQueryData(key)` type-safe.

Thoughts?

**Reply by roblabat** [reply]
Effectively it would provide what we need to get `queryClient.getQueryData(key)` type safe but I'm not very found of the idea of defining a root new QueryClient as we would need to declare here all query options only for type safety, not sure it would be compatible with modularisation and lazy loaded bundles...

I prefer the idea from Vue Inject/provide as defining a shared key is already a good practice I implement even without typings (it simplify query renaming and avoid queryKey miss spelling.

But the downside of this approach is that you have to explicitely define the returned type of the query and cannot infer it for the queryFn return...

Maybe we could have a mixed approach with yours by exporting the key of todoOptions and using it in useQueryClient. It would make something like that:

``` ts
export const todoOptions = queryOptions({
  queryKey: ['todos'],
  queryFn: fetchTodos
})

export const todoKey = todoOptions.queryKey // infered to QueryKey<Todo[]>

useQueryClient().getQueryData(todoKey) // infered to Todo[]
```

with those two mechanism we would have type safety in `queryClient.getQueryData(key)`with type inference of the return type of query and without root definition of the queryClient.

I stay available if you need some implementation examples (I quickly tested the Idea locally and it should work like a charm ^^)

**Reply by roblabat** [reply]
btw I'm not sure we would be able to infer typing from `clientQuery.getqueryData(key)`without the cheat of having a "fake" type on the key like the Vue `InjectKey<T>` type. even if its possible we would for sure have to add a const type to the todoOptions key definition as without it typescript will infer queryKey to be of type string[] and then won't be able to sort the different query definitions.
definition wil have to looks more like this: 

``` ts
const todoOptions = queryOptions({
  queryKey: ['todos'] as const,
  queryFn: fetchTodos
})
```

**Reply by TkDodo** [reply]
> as we would need to declare here all query options only for type safety, not sure it would be compatible with modularisation and lazy loaded bundles...

of course it wouldn't, but that's the trade-off. For type-safety, you need an up-front definition.

But I just realized it won't work with `useQueryClient` anyways, because that hook has no typings and doesn't know which QueryClient it will return, so the proposal is not good.

> Maybe we could have a mixed approach with yours by exporting the key of todoOptions and using it in useQueryClient. 

I think this is promising and could in fact work, but the `queryFn` would need to be required in `queryOptions` (which I don't think it currently is). Do you want to look into that?

**Reply by TkDodo** [reply]
another alternative would be to make an option of `queryClient.getQueryData` that accepts the whole `queryOptions` object:

```
queryClient.getQueryData(todoOptions)
```

and given that `todoOptions` is typed correctly, we could take the types from the `queryFn` passed into it (or unknown if there is no queryFn)

**Comment by TkDodo** [comment]
here's a first poc in that direction, thanks to @Andarist 🙏 

[typescript playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3mgngWwCMAueAVxQEsBHUhXQpCAbgCgMcAHBAESgygAqUAOYAeAQBp4ANQB88ALzwB8AGTwA3i3jwA2vQIBdEtNYBfVi1CRYCAGbkwGCqng0QMHAHkOz1GlFefikAaRAcWQAKJF8XdBItHXdPADEUEkiASkV5AAUYJDwKNBBAvihZViTaTzCcEjrzTITtNxqcNIzshTyCopKy-krW5Jw6kiDBEVE6qUnhixYWZHQsDCRgJB8-dEU2j29Y-0jE-drwkl0AInXNtCvDCRH2zvgsnPh8wuKQADo4NCMABuIEiugAjFIAExSADMhkyTzMmUsK0w8A4SEw2ziaD2oxxx1Oo3GeiumMw90ezwOr3ePU+fR+-xAgIgIJOGKxGDQlyu4KuUiuUMF8CusIe8GRSJRS2s0Dg8AcKCccXgwhAGAAiu1JuIRBrgHV4CAAB4YEAoYB4yZCMTkADWKCQAHcUFJHc63bIoq1+MJDaS7YHwk9mm9gyAjeFdOwuEg7Nh8EZ4AAfMhWkB2CgoKOy+W2JWOHbqzU6g56gTl1IoSZ7T2ulC+6oHUmYGA54S6anhqsvWvlNMZ0DZ3PAJZotYbLF7DXa3XlSK3LZHdC-EnhWUAei3Oh0AD0APwT-xYCk82dlhf8SLnwlrjc4be7vdHpYsHfwAhQMAOl2wa1EEKDg+AoAgIBAE9VngGAoCtQpL3nCtF2uWD4LwB5Mg-F8D2PZZTxguDNjwKFEOrHA9XbTsolQojCkw7C93gN8gA)

**Reply by TkDodo** [reply]
PR is up:
- https://github.com/TanStack/query/pull/6119

**Reply by roblabat** [reply]
I where thinking of a similar solution but little simpler. I don't really understand the interest in adding the Key generic to the type DataTag. I where thinking of a solution like that: [typescript playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3mgngWwCMAueAVxQEsBHUhXQpCAbgCgMcAHBAaRB3gC82DDAooA5gG0AuqxZiMIGADMoYBABEoGKABUo4gDy6AfPBAAPRSmBp4vfgG8AvvBYtQkWAmXkwGClR4GiUcAHkOANQ0Qy0dEwAKJEjA9BJHFnhg2hgcADEUEgSASkEzAAUYJDwKNBBY7SgTViyQ3IcSOL0DBvjWZ2L0zOzQgqLSgQqqmrreppaR9r5Oxv0jLuaWZzlkdCwMJGAkCKj0QUXwlOiEjNacnA74SQAiA6O0Z+kAGmG2-ML4CUyvBKtVaiAAHRwNCMABuIASkgAjF94AAmVEAZmkxR+Ax20SwHCQmBOqTsQj+ZOutwujxexMwH2+v3uY0BEymYLqUJAMIg8Ju8EZGDQJBeSOeqOeaKl8GemM+8AGeOKck80Dg8F8KH8qXg4hAGAAivcusZEsMANbLeBdNYWn6DQG6eAAHzINhAyjEIGAao84E1Pj8pwNRtNoXNukjuQKXXO5CtKCQAHcUJa7qFHphRBIZE6SDG2SgEx7yKAfSg-e5dph4G8SedDSazY0Eo3qegIX8HAGAPT9rJZAB6AH5a4ThSTRc2I22dAkRV20D37n3WIPh-Bx+4WFuCGoranYLZENUONoKAQICBJ3t4DAoDZqnPW1H2y8ny+8J9ivuh2HXcWDrLBvyOPA0TfWMcHNXMxHERIv2fCC-wA7ddyAA)

(solution directly inspired by Vue inject/provide implementation)

What do you think of this solution ?

PS: maybe a better wording for the type DataTag could also be found ?

**Reply by TkDodo** [reply]
yeah this looks a bit easier indeed. Do you want to make a PR to my PR to update it 😅 ?

**Reply by roblabat** [reply]
Yes I'm working on it, I'm just struggling a bit on ts config not allowing unused generics. I come back to you when I'll have something on that.

**Reply by roblabat** [reply]
Here is my PR [https://github.com/TanStack/query/pull/6123](https://github.com/TanStack/query/pull/6123)

**Reply by roblabat** [reply]
There is still the issue about the select method, I think I need a new approve to solve this one (I've got a few ideas to test but not sure it will work).

btw I'm not sure to really understand the functionality behind select. It looks like a way to have something like a map in functional programming to update the result of the query but I would expect from such a functionnality to work between the cache and the return of useQuery but looking at the source it seems like it updates the cache too caching the results of the select function and not result of the queryFn function. With this behaviour using this select like in the test is quite a really bad practice as you could have 2 different queries running using same queryKey with different select and then fighting for the cache... But maybe I had a bad understanding from the source ?

**Reply by TkDodo** [reply]
that's exactly the use-case you want with select: it runs after the queryFn, on observer level; so every instance of useQuery can additionally transform cached data to their needs.

I don't see any issues in your PR with select though ?

**Reply by roblabat** [reply]
Yes it's because I maid a quick fix but it is not really satisfying for me.

The original issue comes from the function refetchInterval that use TData (type of the output of select) as input then by setting just select as typescript doesn't know from the UseQueryOptions type of the options if refetchInterval is set or not it doesn't accept the change it TData as it could break refecthInterval. To make it work in the test I override refetchInterval with undefined then typescript can infer the new TData without conflict. you can see it on line 58 of my PR packages/react-query/src/__tests__/useQuery.types.test.tsx file.

**Reply by roblabat** [reply]
Maybe it's an error but I would expect from refetchInterval to work with the cache and not with the result of select. I tried to see in the source with which value it works but I'm not sure I would need more time reading the source to be sure. If it use the cache it should use the type TQueryFnData and not TData. Solving this type should also solve the select issue without having to use my cheaty quickFix

**Reply by roblabat** [reply]
> that's exactly the use-case you want with select: it runs after the queryFn, on observer level; so every instance of useQuery can additionally transform cached data to their needs.

Then my understanding of the functionality is good but not my understanding of the source as I were understanding the opposite reading the source code... But I only spent few minutes on that so it's probably just a missunderstanding of the source from my side ^^

**Reply by TkDodo** [reply]
> refetchInterval to work with the cache and not with the result of select

refetchInterval gets the result of select passed (TData) because as with select, refetchInterval can be enabled or disabled per observer. It also gets the whole query passed as a second argument though, and there, you have `query.state.data` in the TQueryFnData shape.

**Reply by roblabat** [reply]
Ok that's not how I use Tanstack query as for me there is only one query that rule it's cache and behavior and share an observable (I use a custom hook with the useQuery inside for each query. But I can understand that you could want refetch interval to work with TData in different architectures. I didn't find a clear description of this behaviour of select and refetchInterval in the documentation, maybe I just missed it but if there is not it would be interesting to had one ?

btw With this behaviour I don't have any clear solution for our select typing issue. Maybe we could simply document it as a limitation with the workaround of setting the refetchInterval to undefined until we find a solution. A solution could be to use similar approach to the one used for infering if there is an initial value or not but I don't like that kind of solution as it can become very complicated hard to maintain if you have to follow multiple properties, I have an other one based on conditional typings but I didn't succeed yet in adapting it to our use case. If you are interested My base idea of implementation looks like that [typescript playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3mgngWwCMAueAVxQEsBHUhXQpCAbgCgMcAHBAaRB3gC82DDAooA5gG0AuqxahIsBMnRZgUDFAAqUcQGV8BRiXLVa2Q4zacEAEQ3bdAHi02ANPABqUCLQB8gvCuXPAAZPAA3izw8JLqmjr6lhDSAPwk3r4gLAC+LNYhwSDAAIq0MDi8OE72mgFCtY7iTlUejX75YhggMABmUGAIAEpQAO5lPTgA8hwYFKhoLhMVAGIojYFQKDgeWhtCWss4a+2R0fA0k2vp8AAUAJSCAQAKMEh4FGggS+XH6w5+VgxS4VKokKpA7AgCDgDA3W7xKAkQ6-E4OR4CAJ7BysPIsLo9fqDeBHGZzBY-K7-TSbba7fZBI5ourwEAAD26KGAaHgI3GvzJ83QlNW1Kg9IBZ2BvzBQRspRlfBFf3auPyCmgcHgvXIYHJKAuAtmQsWgoWrI5IC5PL5pONCz8tyQ9vQJDN6Hubo+GCc7rQHgA5CDKnwAwFwhFg7KigrJlVfS60JIg6iUAHpBbOdy7hiXm8Pl8nGJej14KdUmWHPATCgANYoJCjFB+HJyFSYeAYJDAJB+wLBv23KLSuN8EiSABEXZ7aAn0jc52DaxIDye8Fe70+IAAdHA0IwAG4gW6SACMHgATB4AMzSe4LnL3NsLLAcJCYPtCAeJoeLxU4ccJzfTBZ3nP8qRXXN13zLddxAfcICPId4GAjA0EA08Jw8CcLyw+AJ2vOd4EfB8n3ydssAbNZP0NSZB2HWjQTHWIJyolA51yMj8RQbo+gGBAAFUviOIZ4NICAfRRKkGS2HYggZKTRVOBi93EjASA3AtvmxOo1XkcBNQQHUUD1IUyGE34E31Hl2SzHk7Wsx1nWst1E09eA+1sq1swck0ixQEsYEZVNGg8WSAgrP1Jy+GE9SIrzrTuRFkSZMUoOLUsdKgeAKyEkARLEiTlWZcV5MlEg8oKtA1OKsUAhIFAQCPGA5A1JRtV1fV4HEEAMCORoXEdc5a2YmMjnjLQ-AXdzbi0eAAB8yC5EBejEYoyLarVjNM1But6-qHFqhlyHrRtm1ucCmIA4RRAkGRppSkKq0W8hQFWxrgHIl9O27d9Ah6vrfkaW5p17RNtyjPgyIAemhmIYgAPVSL7VBQ980P+-agYcW5UL9CH-xhuH4aRlGOzYzHAcmYG2PxyGcCJ+H4FJ-JYfgAgBlrUZYGzZA8A4DQKAIGEyawGAth7PBKYOzQTwncWuXeOd7hYNmSeRlgKPgBXJYvaXsc0JxMFu8RHUnHWlbvVXicRjWtb3LRfp5IRSAsyYQd+v17jg6qJOtpmWa1mLYWKR2Z0CV38t+X8Ym3OPQb9BcYmDvUVz3BrSEIHoZCgvdtxhCQMAAC23Ls9BEMRxAeTifbU1g1dtoA)

**Reply by TkDodo** [reply]
since we removed the `onSuccess` callback in v5, I think `refetchInterval` is the only one left that gets data transformed via `select` passed. What if we changed that and made the function only get the query passed?

```diff
- refetchInterval: (data, query) => ...
+ refetchInterval: (query) => ...
```

I think that should make the troubles go away, no? It would also be more in line with other callbacks, like the one for `refetchOnWindowFocus`, which is:

```
refetchOnWindowFocus: boolean | "always" | ((query: Query) => boolean | "always")`
```

this too is an observer level property, but it doesn't receive transformed data either ...

**Reply by TkDodo** [reply]
I think I like that idea regardless, we should probably make a separate PR and then see if that fixes the type issues ?

**Reply by TkDodo** [reply]
the change was straight forward, can you try against this PR please:
- https://github.com/TanStack/query/pull/6128

**Reply by TkDodo** [reply]
oh yeah, I could revert the test back to what it was; love it:

https://github.com/TanStack/query/commit/4695de83162575d74b7c622bbd3573776aee21e8

this is a breaking change, and its a bit late given that we are in rc, but I doubt that many people use refetchInterval as a function + select together. If you really need that, you can still run the select function manually against `query.state.data` inside the interval function ...

**Reply by roblabat** [reply]
Yes as refetchInterval is the last callback using TData changing it like in your PR should do the trick. I'll rebase my PR on your branch when I'll have some time ^^

> I think I like that idea regardless, we should probably make a separate PR and then see if that fixes the type issues ?

Yes a separate PR would be better. Currently my main issue with this new approach is type inference failing to infer arguments in callbacks like select, you can see this [here](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3mgngWwCMAueAVxQEsBHUhXQpCAbgCgMcAHBAaRB3gC82DDAooA5gG0AuqxahIsBMnRZgUDFAAqUcQGV8BRiXLVa2Q4zacEAEQ3bdAHi02ANPABqUCLQB8gvCuXPAAZPAA3izw8JLqmjr6lhDSAPwk3r4gLAC+LNYhwSDAAIq0MDi8OE72mgFCtY7iTlUejX75YhggMABmUGAIAEpQAO5lPTgA8hwYFKhoLhMVAGIojYFQKDgeWhtCWss4a+2R0fA0k2vp8AAUAJSCAQAKMEh4FGggS+XH6w5+VgxS4VKokKpA7AgCDgDA3W7xKAkQ6-E4OR4CAJ7BysPIsLo9fqDeBHGZzBY-K7-TSbba7fZBI5ourwEAAD26KGAaHgI3GvzJ83QlNW1Kg9IBZ2BvzBQRspRlfBFf3auPyCmgcHgvXIYHJKAuAtmQsWgoWrI5IC5PL5pONCz8tyQ9vQJDN6Hubo+GCc7rQHgA5CDKnwAwFwhFg7KigrJlVfS60JIg6iUAHpBbOdy7hiXm8Pl8nGJej14KdUmWHPATCgANYoJCjFB+HJyFSYeAYJDAJB+wLBv23KLSuN8EiSABEXZ7aAn0jc52DaxIDye8Fe70+IAAdHA0IwAG4gW6SACMHgATB4AMzSe4LnL3NsLLAcJCYPtCAeJoeLxU4ccJzfTBZ3nP8qRXXN13zLddxAfcICPId4GAjA0EA08Jw8CcLyw+AJ2vOd4EfB8n3ydssAbNZP0NSZB2HWjQTHWIJyolA51yMiWAoqEYT1Gjv31NBfxHJiAJYr4+IwOcF1Ev5ILXDcCx3PdD2PANukwAN73OSTYRXREoMRbcYQkDAAAt4AAeis+BzJ6BBEXgT54AbLBiwc4B4C2LzVMQ+DOyQbztm3ULOM6FBuj6AYEAAVS+I4hng0gIB9FEqQZLYdiCBl0tFU4GL3FKMBIJStxcVVcnVcBNQQHUUD1IUyAS34EyEzMrWzO0hMdZ0hLdRNPXgPt2SzHlupNIsUBLGBGVTRoPCygIKz9Sc9L1IjRs6nkEQcZEmTFKCPNm7EaQreKQES5LUuVZlxRyyUSAuq60GK26xQCEgUBAI8YDkDUlG1XV9XgcQQAwI5GhcR1zlrZiYyOeMtD8BchtuLR4AAHzILkQF6MRijIgGtXqxrUFB8HIYcd6GXIetG2bW5wLEkhMFECQZFR-b5qrbHyFAfHvuAciX0CmdAjBiHfkaW5p17RNtyjPgyJsmIYgAPVSEXVBQ980IlynpYcW5UL9RX-xV2y1c17WOzYg2pcmGW2LNpWcEttX4Bt-JVYIAZa1GWBs2QPAOA0CgCBhW2sBgHz3gdqnNBPCdY65d453uFhVetrXuNF1OezwC8E6NzQnDZsRxEdScC-Tu8s6tjXc54vctG7d9AlIFrJll9u-XuODXtShvPe9nj1u6YA2-FoQu8u34RPgULtzlv1ZN4-S7j3L7SEIHoZCgvcTKtcQLJXpA9BESuHk4wfitYbOm+jje9U77uKluCf+7v1KH8br3UhAA) if you have any idea to solve this I could then try to make a PR for that.

**Comment by roblabat** [comment]
@TkDodo

I made a PR (https://github.com/TanStack/query/pull/6227) to add this new functionality to vue-query as it's currently only available on react-query.

I also investigated a bit more about the issue of the approach described [here](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3mgngWwCMAueAVxQEsBHUhXQpCAbgCgMcAHBAaRB3gC82DDAooA5gG0AuqxahIsBMnRZgUDFAAqUcQGV8BRiXLVa2Q4zacEAEQ3bdAHi02ANPABqUCLQB8gvCuXPAAZPAA3izw8JLqmjr6lhDSAPwk3r4gLAC+LNYhwSDAAIq0MDi8OE72mgFCtY7iTlUejX75YhggMABmUGAIAEpQAO5lPTgA8hwYFKhoLhMVAGIojYFQKDgeWhtCWss4a+2R0fA0k2vp8AAUAJSCAQAKMEh4FGggS+XH6w5+VgxS4VKokKpA7AgCDgDA3W7xKAkQ6-E4OR4CAJ7BysPIsLo9fqDeBHGZzBY-K7-TSbba7fZBI5ourwEAAD26KGAaHgI3GvzJ83QlNW1Kg9IBZ2BvzBQRspRlfBFf3auPyCmgcHgvXIYHJKAuAtmQsWgoWrI5IC5PL5pONCz8tyQ9vQJDN6Hubo+GCc7rQHgA5CDKnwAwFwhFg7KigrJlVfS60JIg6iUAHpBbOdy7hiXm8Pl8nGJej14KdUmWHPATCgANYoJCjFB+HJyFSYeAYJDAJB+wLBv23KLSuN8EiSABEXZ7aAn0jc52DaxIDye8Fe70+IAAdHA0IwAG4gW6SACMHgATB4AMzSe4LnL3NsLLAcJCYPtCAeJoeLxU4ccJzfTBZ3nP8qRXXN13zLddxAfcICPId4GAjA0EA08Jw8CcLyw+AJ2vOd4EfB8n3ydssAbNZP0NSZB2HWjQTHWIJyolA51yMiWAoqEYT1Gjv31NBfxHJiAJYr4+IwOcF1Ev5ILXDcCx3PdD2PANukwAN73OSTYRXREoMRbcYQkDAAAt4AAeis+BzJ6BBEXgT54AbLBiwc4B4C2LzVMQ+DOyQbztm3ULOM6FBuj6AYEAAVS+I4hng0gIB9FEqQZLYdiCBl0tFU4GL3FKMBIJStxcVVcnVcBNQQHUUD1IUyAS34EyEzMrWzO0hMdZ0hLdRNPXgPt2SzHlupNIsUBLGBGVTRoPCygIKz9Sc9L1IjRs6nkEQcZEmTFKCPNm7EaQreKQES5LUuVZlxRyyUSAuq60GK26xQCEgUBAI8YDkDUlG1XV9XgcQQAwI5GhcR1zlrZiYyOeMtD8BchtuLR4AAHzILkQF6MRijIgGtXqxrUFB8HIYcd6GXIetG2bW5wLEkhMFECQZFR-b5qrbHyFAfHvuAciX0CmdAjBiHfkaW5p17RNtyjPgyJsmIYgAPVSEXVBQ980IlynpYcW5UL9RX-xV2y1c17WOzYg2pcmGW2LNpWcEttX4Bt-JVYIAZa1GWBs2QPAOA0CgCBhW2sBgHz3gdqnNBPCdY65d453uFhVetrXuNF1OezwC8E6NzQnDZsRxEdScC-Tu8s6tjXc54vctG7d9AlIFrJll9u-XuODXtShvPe9nj1u6YA2-FoQu8u34RPgULtzlv1ZN4-S7j3L7SEIHoZCgvcTKtcQLJXpA9BESuHk4wfitYbOm+jje9U77uKluCf+7v1KH8br3UhAA) it seems related a typescript inference limitations when working with nested extends. There seems to be no solutions at least in current state of TS, maybe we would be able it work with future releases of TS.

Also my PR on Vue is implemented based on current type for QueryKey based on the first poc from @Andarist so if we want to migrate to my suggestion from PR https://github.com/TanStack/query/pull/6201 it will need an update to make the Vue code working in the new approach. So keep me touch if you want to switch to my type suggestion and I'll make the update.
